File: predecessors/scraper_python/src/scraper/site/site.py
import logging
import base64
import datetime
from typing import List
from requests.sessions import Session
from bs4 import BeautifulSoup
from requests.exceptions import RequestException
from src.storage.storage import StorageUtils
from src.scraper.site.detect_technology import detect
from src.scraper.encoder.encoder import encode_url

logger = logging.getLogger(__name__)

class SiteScraper:
    def __init__(self, job_identifier: str, session: Session = None, storage_utils: StorageUtils = None):
        self.known_urls = []
        self.job_identifier = job_identifier
        self.storage_utils = storage_utils
        self.session = session or Session()

    def __call__(self, base_url: str):
        currrency_code = "ZAR"

        name, image, technology = self.get_site_info(base_url)
        site_id = encode_url(base_url)

        current_datetime = datetime.datetime.now()
        formatted_datetime = current_datetime.strftime("%Y-%m-%d|%H-%M-%S")

        path_prefix = f"site/{self.job_identifier}"
        file_name = f"{path_prefix}/{site_id}_{formatted_datetime}_site.csv"
        items = {
            "id": site_id,
            "name": name.strip() if name else "",
            "image": image.strip() if image else "",
            "currency": currrency_code,
            "technology": technology,
            "scraper_file": f"/{technology}/default.py"
        }
        if self.storage_utils:
            self.storage_utils.write_data_to_csv(file_name, [items])
            return {"file_name": file_name, "site_data": items}
        return items

    def get_site_info(self, url):
        # Send an HTTP GET request to the website and retrieve the HTML content
        response = self.session.get(url)
        html_content = response.content

        # Parse the HTML content using BeautifulSoup
        soup = BeautifulSoup(html_content, "html.parser")

        # Extract the name of the website
        name = soup.title.string
        
        # Extract the image of the website
        image = soup.find("meta", property="og:image")
        image = image["content"] if image else None

        technology = detect(soup)
        # Return the name of the website
        return name, image, technology


    def generate_symbol_map(self):
        # do currencies
        pass
        # import ccy

        # symbol_map = {}
        # currencies = ccy.currencydb()
        # for currency in currencies:
        #     symbol = currencies[currency].symbol
        #     symbol_map[symbol] = currency
        # return symbol_map

File: predecessors/scraper-go/internal/pkg/scrapers/site/site.go
package site

import (
	"fmt"
	"io/ioutil"
	"net/http"
	"strings"
	"time"

	"github.com/exolutiontech/scraper-go/internal/pkg/utils"

	"github.com/PuerkitoBio/goquery"
	"github.com/exolutiontech/scraper-go/internal/pkg/storage"
)

// SiteData is a struct representing the site data
type SiteData struct {
	SiteIdentifier string `json:"site_identifier"`

	Name  string `json:"name"`
	Image string `json:"image"`

	Currency string `json:"currency"`

	Technology string `json:"technology"`
	Scraper    string `json:"scraper"`

	RateLimit string `json:"rate_limit"`

	Url string `json:"url"`
}

// func StructToMap(s SiteData) map[string]string {
// 	return map[string]string{
// 		"id":         s.ID,
// 		"name":       s.Name,
// 		"image":      s.Image,
// 		"currency":   s.Currency,
// 		"technology": s.Technology,
// 		"scraper":    s.Scraper,
// 		"rate_limit": s.RateLimit,
// 	}
// }

type SiteScraper struct {
	Client      *http.Client
	FileStorage storage.FileStorage
}

func New(
	client *http.Client,
	fs storage.FileStorage,
) *SiteScraper {
	return &SiteScraper{
		Client:      client,
		FileStorage: fs,
	}
}

func (s *SiteScraper) ScrapeOne(url string) (map[string]string, error) {
	currencyCode := "ZAR"

	name, image, technology := s.GetSiteInfo(url)
	hostName, err := utils.GetHostName(url)
	if err != nil {
		return nil, err
	}
	hostIdentifier, _, err := utils.StringToIdentifier(url, nil)
	if err != nil {
		return nil, err
	}
	items := []map[string]string{{
		"site_identifier": hostIdentifier,
		"name":            strings.TrimSpace(name),
		"image":           strings.TrimSpace(image),
		"currency":        currencyCode,
		"technology":      technology,
		"rate_limit":      "1/s",
		"scraper":         technology,
		"url":             hostName,
	}}

	if s.FileStorage != nil {

		currentDatetime := time.Now()
		formattedDatetime := currentDatetime.Format("2006-01-02-15-04-05")

		fileName := fmt.Sprintf("site/%s_%s_site.csv", hostName, formattedDatetime)
		err := s.FileStorage.WriteData(items, fileName)
		if err != nil {
			return nil, err
		}
	}
	return items[0], nil
}

func (s *SiteScraper) GetSiteInfo(url string) (string, string, string) {
	// Send an HTTP GET request to the website and retrieve the HTML content
	response, err := s.Client.Get(url)
	if err != nil {
		// Handle the error
	}

	defer response.Body.Close()

	// Read the response body
	body, err := ioutil.ReadAll(response.Body)
	if err != nil {
		// Handle the error
	}

	// Parse the HTML content using goquery
	document, err := goquery.NewDocumentFromReader(strings.NewReader(string(body)))
	if err != nil {
		// Handle the error
	}

	// Extract the name of the website
	name := document.Find("site_name").Text()

	// Extract the image URL from the og:image meta tag
	var image string
	document.Find("meta[property='og:image']").Each(func(i int, selection *goquery.Selection) {
		image, _ = selection.Attr("content")
	})

	// If og:image not found, try finding the first image in div#logo img
	if len(image) == 0 {
		image, _ = document.Find("div#logo img").First().Attr("src")
	}

	// If image is still not found, try finding the first image with link rel="icon"
	if len(image) == 0 {
		document.Find("link[rel='icon']").Each(func(i int, selection *goquery.Selection) {
			iconURL, exists := selection.Attr("href")
			if exists && strings.HasSuffix(iconURL, ".png") || strings.HasSuffix(iconURL, ".jpg") {
				image = iconURL
				return
			}
		})
	}
	technology := Detect(response, body)
	// Return the name, image, and technology of the website
	return name, image, technology
}
